File: cmd/gentle/output.go
/*
 * gogently - a universal toolkit for interpreters
 * Copyright (C) 2021  Michael D Henderson
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published
 * by the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package main

import (
	"fmt"
	"io"
	"os"
)

func SetOption_SUBDIR() {
	SUBDIR = true
}

func Tell(Name string) {
	Told()
	var err error
	if OutFile, err = os.OpenFile(Name, os.O_WRONLY|os.O_CREATE, 0664); err != nil {
		fmt.Printf("cannot open %q: %+v\n", Name)
		os.Exit(1)
	}
}

func TellFile(Name string) {
	if SUBDIR {
		Tell(fmt.Sprintf("_G_/%s", Name))
		return
	}
	Tell(fmt.Sprintf("%s", Name))
}

func TellClauseFile() {
	TellFile(fmt.Sprintf("%s.c", SourceName()))
}

func TellSymbolFile() {
	TellFile(fmt.Sprintf("%s.if", SourceName()))
}

func TellXRefFile() {
	TellFile(fmt.Sprintf("%s.nst", SourceName()))
}

func Told() {
	if OutFile != nil {
		_ = OutFile.Close()
		OutFile = nil
	}
}

const EMIT_CR = false

var OutFile io.WriteCloser
var SUBDIR = false

func doublequote() {
	_, _ = fmt.Fprint(OutFile, "\"")
}

func i(N int) {
	_, _ = fmt.Fprintf(OutFile, "%d", N)
}

func nl() {
	if EMIT_CR {
		_, _ = fmt.Fprintln(OutFile, "\r")
		return
	}
	_, _ = fmt.Fprintln(OutFile, "")
}

func qu_s(Str string) {
	_, _ = fmt.Fprintf(OutFile, "%q", Str)
}

func s(Str string) {
	_, _ = fmt.Fprint(OutFile, s)
}

File: gentle/output.c
/*
   GENTLE 97 CAMPUS EDITION

   COPYRIGHT (C) 1992, 1997. All rights reserved.

   Metarga GmbH, Joachim-Friedrich-Str. 54, D-10711 Berlin

   gentle-97-v-4-1-0
*/


#include <stdio.h>

#define OutBufSize 6000
#define FlushPos 5000


static char OutBuf[OutBufSize];
static char *OutBufPtr;
static FILE *OutFile;
static long OutFileIsOpen = 0;

/*----------------------------------------------------------------------------*/

void Tell(char *Name) {
    Told();
    OutFile = fopen(Name, "w");
    if (OutFile == NULL) {
        char msg[200];
        sprintf(msg, "cannot open %s\n", Name);
        exit(1);
    }
    OutBufPtr = &OutBuf[0];
    OutFileIsOpen = 1;
}

/*----------------------------------------------------------------------------*/

void Told() {
    if (OutFileIsOpen) {
        fwrite(OutBuf, 1, OutBufPtr - &OutBuf[0], OutFile);
        fclose(OutFile);
        OutFileIsOpen = 0;
    }
}

/*----------------------------------------------------------------------------*/

void s(char *Str) {
    while(*Str) {
        *OutBufPtr++ = *Str++;
    }
}

/*----------------------------------------------------------------------------*/

void qu_s(char *Str) {
    *OutBufPtr++ = '\"';

    while(*Str) {
        switch (*Str) {
            case '\\':
                *OutBufPtr++ = '\\';
                *OutBufPtr++ = '\\';
                Str++;
                break;
            case '\n':
                *OutBufPtr++ = '\\';
                *OutBufPtr++ = 'n';
                Str++;
                break;
            case '\"':
                *OutBufPtr++ = '\\';
                *OutBufPtr++ = '"';
                Str++;
                break;
            default:
                *OutBufPtr++ = *Str++;
        }
    }

    *OutBufPtr++ = '\"';
}

/*----------------------------------------------------------------------------*/

void doublequote () {
    s("\"");
}

/*----------------------------------------------------------------------------*/

void i(long N) {
    long butlast;
    long last;
    if (N < 0) {
        *OutBufPtr++ = '-';
        N = - N;
    }
    last = N % 10;
    butlast = N / 10;
    if (butlast > 0) {
        i(butlast);
    }
    *OutBufPtr++ = last + '0';
}

/*----------------------------------------------------------------------------*/

void nl() {
    #ifdef EMIT_CR
        *OutBufPtr++ = '\r';
    #endif
    *OutBufPtr++ = '\n';
    if (OutBufPtr > &OutBuf[FlushPos]) {
        fwrite(OutBuf, 1, OutBufPtr - &OutBuf[0], OutFile);
        OutBufPtr = &OutBuf[0];
    }
}

/*----------------------------------------------------------------------------*/

static long SUBDIR = 0;

/*----------------------------------------------------------------------------*/

void SetOption_SUBDIR() {
    SUBDIR = 1;
}

/*----------------------------------------------------------------------------*/

void TellFile(char *Name) {
    char buf[200];
    if (SUBDIR) {
        sprintf(buf, "_G_/%s", Name);
    } else {
        sprintf(buf, "%s", Name);
    }
    Tell(buf);
}

/*----------------------------------------------------------------------------*/

void TellClauseFile() {
    char name[100];
    extern char *SourceName();
    sprintf(name, "%s.c", SourceName());
    TellFile(name);
}

/*----------------------------------------------------------------------------*/

void TellSymbolFile() {
    char name[100];
    extern char *SourceName();
    sprintf(name, "%s.if", SourceName());
    TellFile(name);
}

/*----------------------------------------------------------------------------*/
void TellXRefFile() {
    char name[100];
    extern char *SourceName();
    sprintf(name, "%s.nst", SourceName());
    TellFile(name);
}

/*----------------------------------------------------------------------------*/
